zarr/storage.py

```diff
@@ -110,7 +115,159 @@ def init_store(store, shape, chunks, dtype=None, compression='default',
     store['attrs'] = json.dumps(dict()).encode('ascii')
 
 
-class DirectoryStore(MutableMapping):
+# backwards compatibility
+init_store = init_array
+
+
+def init_group(store, overwrite=False):
+    """Initialise a group store.
+
+    Parameters
+    ----------
+    store : MutableMapping
+        A mapping that supports string keys and byte sequence values.
+    overwrite : bool, optional
+        If True, erase all data in `store` prior to initialisation.
+
+    """
+
+    # delete any pre-existing items in store
+    if overwrite:
+        store.clear()
+
+    # initialise attributes
+    if 'attrs' not in store:
+        store['attrs'] = json.dumps(dict()).encode('ascii')
+
+
+def check_array(store):
+    return 'meta' in store
+
+
+def check_group(store):
+    return 'meta' not in store and 'attrs' in store
+
+
+class HierarchicalStore(object):
+    """Abstract base class for hierarchical storage."""
+
+    def create_store(self, name):
+        raise NotImplementedError
+
+    def get_store(self, name):
+        raise NotImplementedError
+
+    def require_store(self, name):
+        raise NotImplementedError
+
+    def has_store(self, name):
+        raise NotImplementedError
+
+    def del_store(self, name):
+        raise NotImplementedError
+
+    def stores(self):
+        raise NotImplementedError
+
+
+def ensure_bytes(s):
+    if isinstance(s, binary_type):
+        return s
+    if hasattr(s, 'encode'):
+        return s.encode()
+    if hasattr(s, 'tobytes'):
+        return s.tobytes()
+    if PY2 and hasattr(s, 'tostring'):
+        return s.tostring()
+    return io.BytesIO(s).getvalue()
+
+
+class MemoryStore(MutableMapping, HierarchicalStore):
```

Comment:
Could this be a generic `MutableMappingStore` if the container were passed in instead of hard-coded to be a `dict`?